keps/draft-20180806-namespace-initializer.md

```diff
@@ -0,0 +1,345 @@
+but they still want to enforce the policies in each namespace automatically, including RBAC, ResourceQuota, NetworkPolicy etc.
+They define these policies in a centralized places (e.g. git, corp AD, etc.),
+and use a controller watching namespace creation and populate the predefined policies during namespace creation before the creator gains full permission inside the namespace.
+
+This is the user story demonstrated by the [Namespace Population](https://github.com/kubernetes/community/pull/2177) proposal.
+
+#### Automated External Service Provisioning
+
+The Kubernetes cluster administrator wants to associate namespaces with log analysis system backed by ElasticSearch.
+During the namespace creation, they want to automatically provision ElasticSearch with a new index dedicated for logs sent from this namespace.
+They deploy a controller to perform the provisioning when watch a new namespace creation,
+and they don't want the users to access the namespace before the provisioning is completed.
+
+### Implementation Details
+
+The mechanism is derived from exising [Initializers](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#initializers).
+For more background and initial discussion, please read comments in [Namespace Population KEP](https://github.com/kubernetes/community/pull/2177).
+
+Briefly, namespace initialization involves a few components:
+
+- A built-in admission controller `NamespaceInitializers`
+- A resource of type `NamespaceInitializerConfiguration`
+- Some logic in storage layer to keep the backward-compatible behavior of the API
+- A few namespace initialization controllers: custom controllers which initialize a namespace by creating resources inside the namespace
+
+They work in the process of:
+
+- The namespace initialization controllers register themselves by creating `NamespaceInitializerConfiguration` resources
+- When a namespace is newly created
+  - The `NamespaceInitializers` admission controller aggregates all `NamespaceInitializersConfiguration` resources and generate `spec.Initializers` in this namespace resource
+  - Once the namespace resource is persisted, the following happens in parallel:
+    - Namespace initialization controllers get notified about the newly created namespace,
+      and they start creating resources inside namespace and also update `spec.Initializers` to remove themselves once the initialization is done
+    - The storage layer will wait until `spec.Initializers` becomes empty before returning the response back to client, unless `includeUninitialized` is true in the request.
+
+#### Core Kubernetes Changes
+
+The generic Initializers uses `metadata.Initializers` to record all pending initializations.
+As this is going to be deprecated, for namespaces, `Initializers` is added to `NamespaceSpec` as an _alpha_ field:
```

Comment:
I would actually prefer to leave it in metadata.

I think it'd be OK/interesting to define a ContainerMetadata type which includes this and use it only for namespaces.

Even if namespace is the only container type k8s ever has, I think initializers still need to be understood by the machinery in many places and therefore spec not a good choice.

Reply:
After reading the rest of the proposal, I'm on the fence about this; I'll leave it here as a discussion point. I think the doc could maybe explain why it wasn't chosen in an alternatives section?

Reply:
If we decided to put that in metadata, I will be happy with that. And we can actually keep the existing Initializers and move it forward. I'd be happy to abandon this PR if we can move Initializers forward.

Reply:
> Even if namespace is the only container type k8s ever has, I think initializers still need to be understood by the machinery in many places and therefore spec not a good choice.

The way this is designed, I don't think initialization is a generic concept.  I think it is specific to namespaces as the only scoping resource we have.  I'd rather have metadata without initializers and the field in the spec of the namespace.

Reply:
I'm not quite ready to give up on the idea that we might eventually gain another scoping resource. We could make a ContainerMetadata which is ObjectMetadata (inlined) + initializers; that way we wouldn't impact other resources.

Reply:
@lavalamp IMO ContainerMetadata is a good idea, I will put it in the proposal.

Reply:
In V1 I would much rather have concrete intitializers on namespace aligned with a concrete finalizers field which won’t change for v1, and worry about hypothetical other initializers added as case may come up.

The current initializers mechanism lacks the ability to allow resources to have their own initialization flow, and at the limit initialization is a domain problem (for instance, a job graph) which shouldn’t be reduced to a generic field.